src/util/poly_util.cpp

```diff
@@ -0,0 +1,250 @@
+ * representations. Use with care!
+ */
+template <typename To, typename From>
+To cast_by_string(const From& f)
+{
+  std::stringstream s;
+  s << f;
+  return To(s.str());
+}
+
+Integer to_integer(const poly::Integer& i)
+{
+  const mpz_class& gi = *poly::detail::cast_to_gmp(&i);
+#ifdef CVC4_GMP_IMP
+  return Integer(gi);
+#endif
+#ifdef CVC4_CLN_IMP
+  if (std::numeric_limits<long>::min() <= gi
+      && gi <= std::numeric_limits<long>::max())
+  {
+    return Integer(gi.get_si());
+  }
+  else
+  {
+    return cast_by_string<Integer, poly::Integer>(i);
+  }
+#endif
+}
+Rational to_rational(const poly::Integer& i) { return Rational(to_integer(i)); }
+Rational to_rational(const poly::Rational& r)
+{
+#ifdef CVC4_GMP_IMP
+  return Rational(*poly::detail::cast_to_gmp(&r));
+#endif
+#ifdef CVC4_CLN_IMP
+  return Rational(to_integer(numerator(r)), to_integer(denominator(r)));
+#endif
+}
+Rational to_rational(const poly::DyadicRational& dr)
+{
+  return Rational(to_integer(numerator(dr)), to_integer(denominator(dr)));
+}
+
+poly::Integer to_integer(const Integer& i)
+{
+#ifdef CVC4_GMP_IMP
+  return poly::Integer(i.getValue());
+#endif
+#ifdef CVC4_CLN_IMP
+  if (std::numeric_limits<long>::min() <= i.getValue()
+      && i.getValue() <= std::numeric_limits<long>::max())
+  {
+    return poly::Integer(cln::cl_I_to_long(i.getValue()));
+  }
+  else
+  {
+    return poly::Integer(cast_by_string<mpz_class, Integer>(i));
+  }
+#endif
+}
+std::vector<poly::Integer> to_integer(const std::vector<Integer>& vi)
+{
+  std::vector<poly::Integer> res;
+  for (const auto& i : vi) res.emplace_back(to_integer(i));
+  return res;
+}
+poly::Rational to_rational(const Rational& r)
+{
+#ifdef CVC4_GMP_IMP
+  return poly::Rational(r.getValue());
+#endif
+#ifdef CVC4_CLN_IMP
+  return poly::Rational(to_integer(r.getNumerator()),
+                        to_integer(r.getDenominator()));
+#endif
+}
+
+Maybe<poly::DyadicRational> to_dyadic_rational(const Rational& r)
+{
+  Integer den = r.getDenominator();
+  if (den.isOne())
+  {  // It's an integer anyway.
+    return poly::DyadicRational(to_integer(r.getNumerator()));
+  }
+  unsigned long exp = den.isPow2();
+  if (exp > 0)
+  {
+    // It's a dyadic rational.
+    return div_2exp(poly::DyadicRational(to_integer(r.getNumerator())),
+                    exp - 1);
+  }
+  return Maybe<poly::DyadicRational>();
+}
+Maybe<poly::DyadicRational> to_dyadic_rational(const poly::Rational& r)
+{
+  poly::Integer den = denominator(r);
+  if (den == poly::Integer(1))
+  {  // It's an integer anyway.
+    return poly::DyadicRational(numerator(r));
+  }
+  // Use bit_size as an estimate for the dyadic exponent.
+  unsigned long size = bit_size(den) - 1;
+  if (mul_pow2(poly::Integer(1), size) == den)
+  {
+    // It's a dyadic rational.
+    return div_2exp(poly::DyadicRational(numerator(r)), size);
+  }
+  return Maybe<poly::DyadicRational>();
+}
+
+void approximate_to_dyadic(poly::Rational& r, const poly::Rational& original)
```

Comment:
Note that this method is used to iteratively approximate a rational by a (dyadic) rational. Hence, the argument `r` is both input and output. Would you still prefer `poly::Rational approximate_to_dyadic(const poly::Rational& r, const poly::Rational& original)` even if it is only ever used in the form `r = approximate_to_dyadic(r, original);`?

Reply:
I tend to agree with @aniemetz that it would be cleaner to actually return the result. It is of course a matter of personal taste but I prefer to use parameters for both input/output only if absolutely necessary (e.g. if it would be much less efficient).